File: steps/python/adoption_followup_event.step.py
# steps/python/adoption_followup_event.step.py
import asyncio

config = {
    "type": "event",
    "name": "PyAdoptionFollowup",
    "subscribes": ["py.adoption.approved"],
    "emits": [],
    "flows": ["pets"]
}

async def handler(event, ctx=None):
    logger = getattr(ctx, 'logger', None) if ctx else None
    
    try:
        import sys
        import os
        sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
        from services import pet_store
    except ImportError:
        return {"success": False, "message": "Import error"}
    
    # Try both event.data and direct event access
    event_data = event.get("data", event)
    pet_id = event_data.get("petId")
    application_id = event_data.get("applicationId")
    
    if not pet_id:
        print("❌ No pet ID provided in adoption event")
        return {"success": False, "message": "Missing pet ID"}
    
    pet = pet_store.get(pet_id)
    if not pet:
        print(f"❌ Pet with ID {pet_id} not found")
        return {"success": False, "message": "Pet not found"}
    
    # Log to both console and structured logger
    print(f"📧 Sending adoption follow-up for {pet['name']}")
    if logger:
        logger.info("Adoption follow-up sent", {
            "petId": pet_id,
            "petName": pet["name"],
            "applicationId": application_id,
            "species": pet["species"]
        })
    
    # Simulate sending follow-up email/notification
    await asyncio.sleep(0.1)
    
    return {
        "success": True,
        "message": f"Follow-up sent for {pet['name']}'s adoption",
        "petName": pet["name"],
        "applicationId": application_id
    }

File: steps/typescript/adoption.followup.event.step.ts
// steps/typescript/adoption.followup.event.step.ts
import { TSStore } from './ts-store';

export const config = {
  type: 'event',
  name: 'TsAdoptionFollowup',
  subscribes: ['ts.adoption.approved'],
  emits: [],
  flows: ['adoptions']
};

export const handler = async (event: any, context?: any) => {
  const logger = context?.logger;

  // Try both event.data and direct event access
  const eventData = event.data || event;
  const { petId, applicationId } = eventData;

  if (!petId) {
    console.log('❌ No pet ID provided in adoption event');
    return { success: false, message: 'Missing pet ID' };
  }

  const pet = TSStore.get(petId);
  if (!pet) {
    console.log(`❌ Pet with ID ${petId} not found`);
    return { success: false, message: 'Pet not found' };
  }

  // Log to both console and structured logger
  console.log(`📧 Sending adoption follow-up for ${pet.name}`);
  if (logger) {
    logger.info('Adoption follow-up sent', {
      petId,
      petName: pet.name,
      applicationId,
      species: pet.species
    });
  }

  // Simulate sending follow-up email/notification
  await new Promise(resolve => setTimeout(resolve, 100));

  return {
    success: true,
    message: `Follow-up sent for ${pet.name}'s adoption`,
    petName: pet.name,
    applicationId
  };
};
